Slm spatial calibration

fthouin opened the pull request:
The Spatial calibration tab of the GUI as well underlying measurement routines are working! They are currently configured to run with demo devices.

You can try it out by performing a vertical beam calibration from the GUI. The manual has been updated to explain how to do that. Try launching a measurement, adding delimiter and setting the beam parameters.

You can also try the code out by performing a spectral calibration. The manual has been updated to explain how it works. Try launching a measurement, reducing the wavelength range, performing a fit with various polynomial degrees and updating the calibration.

These can be performed in any sequence.

fthouin commented:
@SDaneau @MathieuDesmarais could you guys confirm that it works on the lab computer and check that the code makes sense and is clean?

SDaneau commented:
@fthouin, I got this error trying to run main.py : 

Traceback (most recent call last):
  File "C:\Users\colbe\OneDrive\Documents\GitHub\colberto\src\main.py", line 27, in <module>
    from measurements.MeasurementClasses import AcquireMeasurement,RunMeasurement,BackgroundMeasurement, ViewMeasurement, KineticMeasurement
ImportError: cannot import name 'KineticMeasurement' from 'measurements.MeasurementClasses' (C:\Users\colbe\OneDrive\Documents\GitHub\colberto\src\measurements\MeasurementClasses.py)

fthouin commented:
@SDaneau thanks for running it. I fixed this bug and pulled the latest state of  `dev` into this branch

I managed to perform the following on a virtual Windows machine.

1. **Run a vertical beam calibration measurement.** Click on measure in the Vertical SLM calibration tab and see the steps appear. You can change the spectrometer integration time to make this a little faster.
![Screenshot from 2025-05-06 18-05-51](https://github.com/user-attachments/assets/27c25fed-614f-46f4-911f-ff8974d57c1a)
2. **Place the beam delimiters** and see the markers appear on that same plot
![Screenshot from 2025-05-06 18-06-34](https://github.com/user-attachments/assets/f3056bf8-44d6-4bee-9e83-c08e3a2a39d0)
3. **Press Assign beams and period.** So far, all changes are internal and are not accessible by the user
4. **Run a spectral SLM calibration.** Press measure and watch the 2D plot grow with a very fine line. You can change the Y axis to see it is indeed a line. Aliasing make it hard to see otherwise. 
5. Fit the data by pressing the Fit button. You should see the residual appear on the rightmost plot 
![Screenshot from 2025-05-06 18-10-24](https://github.com/user-attachments/assets/6acbcbd5-0b30-4ecb-88b1-8d15c2ff7d34)
6. Change the shortest and longest fitting wavelength to change the spectral region considered in the fit. The plots should be update automatically
7. **Press Assign calibration.**  So far, all changes are internal and are not accessible by the user

Some notable changes:
- I deactivated the automatic spectrum dumping of DataHandling  as it was giving me errors. Now the memory of the datahandler is simply cleared
- I modified the path to the SLM DLLs and it fixed the bug I was having with WinError87 [to fix this issue](https://github.com/SilvaScience/colberto/issues/42)

Lastly, it seems like the SLM image is not being updated either here. I suggest we disregard this problem in the context of this commit as it is probably linked to [an on-going debugging](https://github.com/SilvaScience/colberto/pull/39) tha probably stems from the dev branch itself.

SDaneau commented:
I propose a modification of the Sim.Meadowlark_optics.py to make it works on all platform. The first lines up to  "success = ... (line 24)" should be replaced by : 

import sys
from pathlib import Path
import logging
logger = logging.getLogger(__name__)
import datetime

if sys.platform == "win32":
    import ctypes
    from ctypes import byref, c_int

    try:
        awareness = c_int()
        ctypes.windll.shcore.GetProcessDpiAwareness(0, byref(awareness))
        print("Windows DPI awareness level:", awareness.value)
        ctypes.windll.shcore.SetProcessDpiAwareness(2)
    except Exception:
        try:
            ctypes.windll.user32.SetProcessDPIAware()
        except Exception:
            print("DPI awareness setting failed.")
else:
    print("Skipping DPI settings — this is not Windows.")

fthouin commented:
Hi Simon, 
Thanks for the suggestion but I recommend we develop on the environment we will deploy the software in. This will make bugs less of a nightmare to hunt down.

Developpers not working on a Windows platform should run the code on  a Vritual machine. This is what I am doing and it is working pretty well.

fthouin commented:
I am also noticing the program is very slow when using the demo version of the SLM. Something is horribly wrong with the SLM driver or our use of it in the code.

fthouin commented:
@KatieKoch2 @SDaneau I think I have solved the issue and was able to make the test mentioned above as well as observe the correct pattern  You can make the tests mentioned above either using the SLM_Meadowlark demo, the SLM Meadowlark with a device or using a mockup of the SLM (SLMDemo.py) (triggered by uncommenting `#raise Exception('DEMO')` at line 78 in main.py).

You should see the following in the SLM Tab of the UI when performing a vertical calibration scan
![Screenshot from 2025-05-07 18-21-24](https://github.com/user-attachments/assets/f4c4a957-0675-430b-9b47-939652b6fd6f)
and the following on the SLM/External monitor if not using the SLMDemo file
![Screenshot from 2025-05-07 18-21-39](https://github.com/user-attachments/assets/5ef0c382-42e0-4329-8e2d-72be732928be)

You should also see the following when performing the spectral calibration scan on the SLM tab
![Screenshot from 2025-05-07 19-00-43](https://github.com/user-attachments/assets/e273102d-de6f-48b8-833b-60e3f9faf5df)
and the following on the external monitor
![Screenshot from 2025-05-07 19-01-08](https://github.com/user-attachments/assets/0ffa9e1e-7c91-4684-9fd8-e11910004999)

You will see the display time is much faster than the previous version of the SLM driver.

Finally, I tried running these with very short integration time to speed up the measurements. It seems like below 100ms the measurements don't get much faster, even when using the SLMDemo. Since the latter does not use the DLL from MeadowLark, I suspect the `beams` module is the limiting factor.

I tried this on a virtual machine. Let me know if all of these work on the lab computers. If so, we could pull these changes into [the other pull request](https://github.com/SilvaScience/colberto/pull/39).

SDaneau commented:
Everything is working fine, I can see the phase pattern on the SLM while doing the vertical calibration. Great job @fthouin. This Pull request can be merged. I'll try the calibration procedure with the greyscale.

fthouin commented:
Itssssss aaa mmeeerrgggggeeee!
